torch/optim/optimizer.py

```diff
@@ -164,18 +169,45 @@ def zero_grad(self):
                     p.grad.detach_()
                     p.grad.zero_()
 
-    def step(self, closure):
-        r"""Performs a single optimization step (parameter update).
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step (parameter update).
 
         Arguments:
-            closure (callable): A closure that reevaluates the model and
-                returns the loss. Optional for most optimizers.
+            closure (callable, optional): A closure that reevaluates the model
+            and returns the loss. Optional for most optimizers.
 
         .. note::
             Unless otherwise specified, this function should not modify the
             ``.grad`` field of the parameters.
         """
-        raise NotImplementedError
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            lr = group['lr']
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+
+                if p.grad.is_sparse:
```

Comment:
Since `OptimizerW.step` puts the weight decay in here, it could be a good idea to mirror and add the L2 weight decay here too. But then not all optimizers support it. Thoughts?

Reply:
L2 weight decay is actually not really part of the optimization algorithms, but rather a shortcut for a regularization term in the loss. I think it should be safe to move that into `Optimizer.step`.

Reply:
Since L2 weight decay is actually a short-cut for an extra loss term, I update the `grad` field in the `step` function as if it would have been regularised by an extra loss term. This makes the refactoring easier and at least to me it seems reasonable enough not to bother. I changed the documentation note correspondingly. Any objections from your side?

Reply:
Good point. Modifying the grad field in place is BC-breaking, so we should avoid doing that in this.

We could pass the gradient to `get_update` and `get_sparse_update` (e.g. `get_update(par, par.grad, ...)` and `get_sparse_update(par, par.grad, ...)`). I see `grad = par.grad` as a first line anyway in most update function. With regularization, we could pass `p.grad.add(weight_decay, p)` instead of just `p.grad`. Thoughts?

Otherwise, we could just keep the gradient inside the update function as you are currently doing.

Reply:
Is it BC-breaking? There were some optimizers that changed the gradients in-place when I originally started working on this (if I am not mistaken). I believe that this was fixed with the pull request that removed all `.data` in the optimizers. 

Assuming that this pull request is included before the next release, I would not really consider it backward incompatible (at least not more than the `.data` pull request).

I originally went for passing the gradients to `get_update` as you suggested, but it is somehow very clumsy passing the parameter **and** its gradient separately (and since the optimizer state is only accessible through the parameter, it seems very hard not to pass the parameter to be updated).

Reply:
Hi,
The point of this change was to make sure all the optimizers have a consistent behavior wrt gradient use. In particular, some users measure gradient norm after the optimizer step.
So that would be a regression if we don't have this consistent behavior anymore.

Reply:
I second that: gradient should not be modified in place, and it should not have happened.

Reply:
Behaviour would still be consistent if all optimisers do the modification in place. It could also be argued that L2 regularisation should appear in the gradient, since the implementation in the optimiser it is essentially a shortcut to adding an L2 loss term.

I would argue that the L2 weight decay should appear in the gradients for things to be correct/consistent. After all, you would get different gradients if you implement the L2 regularisation explicitly by adding a loss term.

Reply:
Yes you definitely can argue both ways.
The reason why "not changing" was chosen mainly for its simplicity.
It is that it is much clearer to explain and easier to be sure that third party optimizers do follow this guideline.
It can be more complex to explain that if the optimizer is doing regularization or any modification of the loss function in an implicit way, then the .grad field of the parameters should be updated to reflect the final loss function that corresponds to the gradients.
And for more complex optimizers like LBFGS. Does this second case even make sense? What should this one do?